Handle text_alignment in Milestone

kajuberdut opened the pull request:
This change addresses issue #75 by shifting text 1/2 it's length left or right when "left"/"right" are passed as text_alignment.

kajuberdut commented:
This is a very simple version of aligning Milestone text. I think a better version would be to allow pixels or % as additional arguments like `text_alignment="left:100"` for 100 pixels to the left or `text_alignment="left:50%"` for 50% of text length to the left. If this sounds good I can update my pull request to implement these.

felix-seifert reviewed (commented):
I know too little about the implementation of Roadmapper's functionality, and I highly doubt that we have a useful example for unit tests. However, I highly recommend that when you modify a feature, you write a simple unit test that ensures this functionality works as expected. Besides explaining your actual code changes, such tests ensure that your functionality does not regress with future changes.

kajuberdut commented:
> I know too little about the implementation of Roadmapper's functionality, and I highly doubt that we have a useful example for unit tests. However, I highly recommend that when you modify a feature, you write a simple unit test that ensures this functionality works as expected. Besides explaining your actual code changes, such tests ensure that your functionality does not regress with future changes.

I agree that in general unit tests are a good idea but in this case I haven't any idea how to make a useful assertion about this feature and don't want to write a meaningless test for the sake of having a test. If you can suggest a way to assert that a text has been offset I am happy to write a test.

kajuberdut commented:
``> This is a very simple version of aligning Milestone text. I think a better version would be to allow pixels or % as additional arguments like `text_alignment="left:100"` for 100 pixels to the left or `text_alignment="left:50%"` for 50% of text length to the left. If this sounds good I can update my pull request to implement these.

For my own use I wanted the ability to do alignment by percent or pixel so I implemented it. If you wold rather see this code (or this code with additional requested changes) in the pull request please let me know:

```python
text_width, _ = painter.get_text_dimension(text=self.text, font=self.font, font_size=self.font_size)

# Text is already "centered" if we change nothing
# For "left" and "right", move text_x by 1/2 text_width.
if ":" in self.text_alignment:
    alignment, modifier = self.text_alignment.split(":")
else:
    alignment = self.text_alignment
    modifier = None

if modifier is None:
    offset = .5 * text_width
elif "%" in modifier:
    offset = (int(modifier.strip("%")) / 100) * text_width
else:
    offset = int(modifier)

if alignment == "right":
    self.text_x = self.text_x + offset
elif alignment == "left":
    self.text_x = self.text_x - offset
```

felix-seifert commented:
I had a look at your edited class and method. Testing this method is indeed slightly weird as the method does not really have an output but generates a file.

My suggestion is to create an integration test for this. We already have [quite generic tests](https://github.com/csgoh/roadmapper/blob/main/src/tests/compare_generated_roadmaps_test.py#L80-L95) which basically compare the generated roadmaps with how the roadmap should look. 

Your task would be to create the code for a roadmap similar to [ColourTheme](https://github.com/csgoh/roadmapper/blob/main/src/tests/roadmap_generators/colour_theme.py). Then, add the class of your roadmap to the same lists as `ColourTheme` and `ColourThemeExtensive`. Please pay attention to giving appropriate names.

After you commited your changes, we can help you to generate the example roadmaps on the GitHub servers. Alternatively, you execute [the workflow to generate the roadmaps](https://github.com/csgoh/roadmapper/blob/main/.github/workflows/generate_examples.yaml) from your personal account.

csgoh commented:
Hi,

Sorry for the late response, as I am currently on holiday.

Thanks for the feedback and improvement suggestions. I really appreciate it. 

@kajuberdut It is indeed a very simple text alignment implementation. I looked at your suggestion, and I think it is a good idea. However, as @felix-seifert was saying, it would be great if you could write the corresponding test cases. To be honest, I am pretty bad at writing test cases and would definitely need some help.

CS

kajuberdut commented:
I'll implement the test case over the coming weekend. Thanks!

kajuberdut commented:
Ok, as of today I have re-worked it to allow the options I mentioned above like "left:100" for 100 units left, or "left:75%" for 75% of text width to the left. This change means there is some logic for unit tests so I also added unit tests in: src/tests/test_milstone.py

I also added a src/tests/roadmap_generators/milestone_text_alignment.py to generate a roadmap with examples of the text alignment in milsetones and added it to the same lists that ColourTheme was in.

The unit tests are all passing:
![image](https://github.com/csgoh/roadmapper/assets/8014746/ec3f25db-3319-4a2e-bc77-22d382bf9387)


@felix-seifert I will now look into running the workflows or you are welcome to jump in on those. Thank you.

kajuberdut commented:
Ok, it took a few iterations but I have both workflows passing in my fork: https://github.com/kajuberdut/roadmapper/actions

@felix-seifert or @csgoh if either of you could kick off the workflows when convenient or let me know if I need to do something else first that would be great. Thanks.

felix-seifert commented on src/roadmapper/milestone.py:
```diff
@@ -59,6 +58,38 @@ def draw(self, painter: Painter) -> None:
+                text=self.text, font=self.font, font_size=self.font_size
+            )
+
+            if ":" in alignment:  # Split if ":" in str
+                alignment, modifier = alignment.split(":")
+
+                if "%" in modifier:  # If "%" treat as percent of text_width
+                    offset = (int(modifier.strip("%")) / 100) * text_width
+                else:  # Treat as units
+                    offset = int(modifier)
+            else:  # If no ":", default to half of text width
+                offset = 0.5 * text_width
```
Considering the ideal of clean code, we do not really want nested ifs. They are just more complex to understand and can cause failures when refactoring code. If you wanted to stick to the current solution, I would simply create submethods for this block. This makes it easier to know what this block does without having to think about the details. Additionally, method names are like the better comments: you have description text which has to be refactored if you change something, unlike comments which tend to be forgotten.

Without forcing you to stick to my suggestion, I could think of something like the following 
```
def handle_correct_cases_of(alignment):
    if alignment is None or alignment == "centre":
        return  # Text is already "centre" if we change nothing

    offset = get_offset_from(alignment)
    apply_offset_for(alignment, offset)


def get_offset_from(alignment):
    # Calculat correct offset value
    # Keep in mind that a `return` can eliminate the need for an else/elif
    # if something is x:
    #     return x
    # no_need_for_else()
    # return 5


def apply_offset_for(alignment, offset):
    # Apply offset to self.text_x
```

kajuberdut replied:
This is good feedback and considering that either the parsing of more complex alignments makes sense to add in multiple places, or it makes no sense to add it, I have refactored Alignment it out to it's own class.

felix-seifert commented on src/roadmapper/milestone.py:
```diff
@@ -59,6 +58,38 @@ def draw(self, painter: Painter) -> None:
                 self.fill_colour,
             )
+
+        alignment = (
+            self.text_alignment.lower() if self.text_alignment is not None else None
+        )
```
It would be nice if we caught wrong alignment cases already here. Then, we do not have to think of these wrong cases later on. This means you should move the `else` from below to this place.

felix-seifert commented on src/roadmapper/milestone.py:
```diff
@@ -59,6 +58,38 @@ def draw(self, painter: Painter) -> None:
+            self.text_alignment.lower() if self.text_alignment is not None else None
+        )
+
+        if alignment is None or alignment == "centre":
```
Do you also want to consider the case of `centre:offset_value`?

kajuberdut replied:
This is still a good comment and I'll have to look at how it applies now that I have moved the logic for parsing alignment out to it's own class.

kajuberdut replied:
I added  validation for this since it seems bad to silently ignore the input of an offset with a center alignment.

https://github.com/csgoh/roadmapper/commit/a6af49d308528d91f7f478866adc919beee2d7fc

felix-seifert commented on src/roadmapper/milestone.py:
```diff
@@ -59,6 +58,38 @@ def draw(self, painter: Painter) -> None:
+            self.text_alignment.lower() if self.text_alignment is not None else None
+        )
+
+        if alignment is None or alignment == "centre":
```
Should we consider both spellings `centre` and `center`? (This would then also have to be considered in the error message in the last else block.)

kajuberdut replied:
In my most recent commit the Alignment class handles these two being aliases by using an Enum subclass.

felix-seifert commented on src/tests/test_milestone.py:
```diff
@@ -0,0 +1,101 @@
+    return MockPainter()
+
+
+# Text test cases
```
Is this comment needed?

felix-seifert reviewed (commented):
I love that you add test cases. :heart: Especially that you also discovered this opportunity for unit tests. :heart: In general, I like that you contribute to roadmapper and want to improve it.

The comments I left are not really about missing functionality but more about the code readability. Theoretically, we could just merge your changes. However, we should consider the long-term quality of roadmapper and embrace the possibility of learning.

kajuberdut commented:
In my update just now I refactored alignment out to it's own class so that these offset by units / percent etc options can be easily used in other items that have alignment (however, percent is a bit of a problem since in other cases it is the percent of the bounding object instead of the percent of the text.)

I have added unit tests for the new class and updated Milestone to use the Alignment class to handle most of the logic for alignment. However, I have failing tests and also need to address @felix-seifert's comment about updating the documentation so this is not ready for review again. Also I have run out of weekend so this might have to wait a week for me to complete it. Thanks for the useful feedback so far.

kajuberdut commented on src/roadmapper/milestone.py:
```diff
@@ -59,6 +58,38 @@ def draw(self, painter: Painter) -> None:
+            if alignment == "right":
+                self.text_x += offset
+            elif alignment == "left":
+                self.text_x -= offset
```
Ok. This seems to be the last comment standing however:
1. There do not seem to be code examples in the code anywhere else, there are for the most part no docstrings, so it feels a bit odd to add a docstring with a code example.
2. On the other hand, I can't see any way to include an update to the wiki in this pull request.

I think if this pull request is satisfactory, the next step would be to update the other objects that allow alignment to use the new Alignment class and to offer % and unit offset options. Then the wiki should be updated with a description of that feature.

kajuberdut commented:
I had some time so I fixed the bug I had introduced that was failing unit tests and I also added a validation to avoid the impossible case of center + offset. I think this is ready to review and all tests are passing.
